File: speckit-chat-plugin/src/main/kotlin/com/speckit/plugin/service/DiscoveryService.kt
package com.speckit.plugin.service

import com.speckit.plugin.model.TableRow

/**
 * Pure parsing and serialization logic for discovery.md files.
 * No Swing dependencies.
 */
object DiscoveryService {

    /** Strip YAML front-matter (---...---) from template content. */
    fun extractBody(content: String): String {
        val match = Regex("^---\\s*\\n.*?\\n---\\s*\\n?", RegexOption.DOT_MATCHES_ALL).find(content) ?: return content
        return content.substring(match.range.last + 1)
    }

    /** Parse `## Category` headings and `- Attribute = Answer` lines into rows. */
    fun parseDiscovery(content: String): List<TableRow> {
        val rows = mutableListOf<TableRow>()
        var currentCategory = ""
        for (line in content.lines()) {
            val trimmed = line.trim()
            if (trimmed.isBlank()) continue

            if (trimmed.startsWith("## ")) {
                currentCategory = trimmed.removePrefix("## ").trim()
                continue
            }

            if (trimmed.startsWith("- ") && currentCategory.isNotEmpty()) {
                val eqIdx = trimmed.indexOf('=')
                if (eqIdx > 0) {
                    val attribute = trimmed.substring(2, eqIdx).trim()
                    val answer = trimmed.substring(eqIdx + 1).trim()
                    rows.add(TableRow(currentCategory, attribute, answer))
                }
            }
        }
        return rows
    }

    /** Serialize category data back to markdown format. */
    fun serializeToMarkdown(categories: List<Pair<String, List<Pair<String, String>>>>): String {
        val sb = StringBuilder()
        var first = true
        for ((category, attributes) in categories) {
            if (!first) sb.appendLine()
            first = false
            sb.appendLine("## $category")
            for ((attribute, answer) in attributes) {
                sb.appendLine("- $attribute = $answer")
            }
        }
        return sb.toString()
    }
}

File: speckit-chat-vscode/src/service/discovery-service.ts
import { TableRow } from '../model/discovery-models';

/**
 * Pure parsing and serialization logic for discovery.md files.
 */
export function extractBody(content: string): string {
    const match = content.match(/^---\s*\n[\s\S]*?\n---\s*\n?/);
    if (!match) { return content; }
    return content.substring(match[0].length);
}

export function parseDiscovery(content: string): TableRow[] {
    const rows: TableRow[] = [];
    let currentCategory = '';
    for (const line of content.split('\n')) {
        const trimmed = line.trim();
        if (!trimmed) { continue; }
        if (trimmed.startsWith('## ')) {
            currentCategory = trimmed.substring(3).trim();
            continue;
        }
        if (trimmed.startsWith('- ') && currentCategory) {
            const eqIdx = trimmed.indexOf('=');
            if (eqIdx > 0) {
                const attribute = trimmed.substring(2, eqIdx).trim();
                const answer = trimmed.substring(eqIdx + 1).trim();
                rows.push({ category: currentCategory, attribute, answer });
            }
        }
    }
    return rows;
}

export function serializeToMarkdown(
    categories: Array<{ category: string; attributes: Array<{ attribute: string; answer: string }> }>,
): string {
    const parts: string[] = [];
    for (const { category, attributes } of categories) {
        if (parts.length > 0) { parts.push(''); }
        parts.push(`## ${category}`);
        for (const { attribute, answer } of attributes) {
            parts.push(`- ${attribute} = ${answer}`);
        }
    }
    return parts.join('\n') + '\n';
}
